docs/specs/code/TRANSACTION_MEMO.md

```diff
@@ -0,0 +1,87 @@
+
+# Background
+
+### Components and UI
+A new container was added: `/containers/settings/categories/AccountsSettingsPage.js`.
+It renders an AccountsSettings component.
+
+A new component was added: `/components/settings/categories/AccountsSettings.js`.
+Here user can link his external storage account (only Dropbox for now) to Yoroi.
+
+A new route was added for the AccountsSettingsPage.
+
+A new component was added: `/components/wallet/send/WalletCloseMemoDialog.js`.
+It serves for getting a confirmation if user wants to close a memo textarea when it's not empty.
+
+A new textarea was added to `/components/wallet/send/WalletSendForm.js`
+
+### Flow
+
+The user flow is the following:
+1. User is going to make a transaction. He is looking at `/components/wallet/send/WalletSendForm.js`.
+2. The memo textarea is closed by default. User can click "Add memo" label.
+3. If user clicked "Add memo", he or she can see the following:
+   - If user haven't synced his external storage account yet, he or she will see the text "In order to attach private memos to   transactions you need to link Dropbox account." and a link to AccountsSettingsPage.
+     - If user clicks the link, he or she will be redirected to AccountsSettingsPage. Here user can click "Authorize Yoroi in    Dropbox". If user does, he or she will be asked to authorize Yoroi in Dropbox. After that user will be redirected to a    home page.
+   - If user have already synced his external storage account, he or she will see an opened textarea where the user can enter    the memo.
+4. If user clicks "Add memo" again:
+   - If the textarea is empty, it will be just closed.
+   - If it isn't empty, user will see the `/components/wallet/send/WalletCloseMemoDialog.js` dialog window with the text "Remove memo?" and two buttons, "Yes" and "No".
+     - If user clicks "No", the dialog closes and the memo textarea remain opened and have as content user's memo.
+     - If user clicks "Yes", the dialog window closes and the textarea become cleared and closes too.
+5. If user have entered the memo and makes a transaction, memo will be saved to localStorage and also to user's external storage.
+6. When user goes to his summary - `/containers/wallet/WalletSummaryPage.js` and clicks on the transaction to expand it, he or she will see his memo at the very bottom of Transaction item (if memo exists).
+7. When user links his external storage account to Yoroi, we check if there are some memos in his external storage. If there are, we download a content of each memo file via Dropbox API, read the content with FileReader object, then we save these memos to localStorage. In this way we can make user be able to install Yoroi on the another device, link his or her external storage account and get user's memos so they can be displayed in WalletSummary.
```

Comment:
I don't think we can only check this when they link their account as if the user already has 2 wallets (A & B) on 2 different computers, making a memo on computer A will not make it show up on computer B.

I think we have to check either periodically or every time a new transaction is detected (not created by our wallet).

Reply:
Sorry, I can't get the idea: the user can have only one wallet per instance, so why we need to have his memos for wallet A when he uses the wallet B? Each wallet has it's own summary, so the data for wallet A won't be displayed for wallet B.

Reply:
Oh sorry my bad. I meant same wallet on 2 different computers

Reply:
Aha, you're right. To check periodically is easier, but more neat solution as for me would be to check every time a new transaction is detected. I think we need to use something like websockets, but all the suggestions are welcome.

Reply:
Latest discussion with @JizKadasha:
- When our local wallet is not active - we will locally store decrypted copy of the file and some "remote file version ID" (Which is either latest known upload timestamp, or hash. Hash is better, cuz ensures content-equality.) and on every event `E` we will query remote to compare local version ID with current version ID, and if those don't match - replace local copy with remote (download and decrypt). Where `E` might be either a complex event detection (e.g. when we see new transaction where we are the sender, but we received it from outside - this means there's some background action with another copy of the same walet), or just on every transaction history update, or some time-preiod. Keeping remote file version ID makes check fairly lightweight.
- When we add new memo - we add it to local copy of the file, then encrypt the file, and then upload it to remote, at this point Dropbox API will potentially save our file as "conflicting copy", at which point we need to download current remote file, merge them and upload again. Background wallet B then assumed to eventually "background sync" with the new merged file, by checking that its local version ID does not match.

Reply:
@JizKadasha, plz update the flow according to these latest discussions and then it will be 👌